src/Application/GameWindowHandler.cpp

```diff
@@ -29,9 +33,84 @@ using ApplicationIoc = Application::IocContainer;
 using Application::GameWindowHandler;
 using Io::InputAction;
 
+static char PlatformKeyToChar(PlatformKey key, PlatformModifiers mods) {
+    if (key >= PlatformKey::Digit0 && key <= PlatformKey::Digit9) {
+        return std::to_underlying(key) - std::to_underlying(PlatformKey::Digit0) + '0';
+    } else if (key >= PlatformKey::A && key <= PlatformKey::Z) {
+        if (mods & PlatformModifier::Shift) {
+            return std::to_underlying(key) - std::to_underlying(PlatformKey::A) + 'A';
+        } else {
+            return std::to_underlying(key) - std::to_underlying(PlatformKey::A) + 'a';
+        }
+    }
+
+    return 0;
+}
+
 
 GameWindowHandler::GameWindowHandler() {
     this->mouse = EngineIoc::ResolveMouse();
+    this->keyboardController_ = std::make_unique<GameKeyboardController>();
+}
+
+void GameWindowHandler::UpdateWindowFromConfig() {
+    std::vector<Recti> displays = platform->DisplayGeometries();
+
+    Pointi pos = {engine->config->window.PositionX.Get(), engine->config->window.PositionY.Get()};
+    Sizei size = {engine->config->window.Width.Get(), engine->config->window.Height.Get()};
+
+    Recti displayRect;
+    int display = engine->config->window.Display.Get();
+    if (display > 0 && display < displays.size()) {
+        displayRect = displays[display];
+    } else if (!displays.empty()) {
+        displayRect = displays[0];
+    }
+
+    if (engine->config->window.Fullscreen.Get()) {
+        pos = displayRect.TopLeft();
+    } else if (Recti(Pointi(), displayRect.Size()).Contains(pos)) {
+        pos += displayRect.TopLeft();
+    } else if (displayRect != Recti()) {
+        pos = displayRect.Center() - Pointi(size.w, size.h) / 2;
+    } else {
+        pos = Pointi(0, 0); // DisplayGeometries() failed so we just give up.
+    }
+
+    window->SetPosition(pos);
+    window->Resize(size);
+    window->SetTitle(engine->config->window.Title.Get());
+    window->SetBorderless(engine->config->window.Borderless.Get());
+    window->SetGrabsMouse(engine->config->window.MouseGrab.Get());
+    window->SetFullscreen(engine->config->window.Fullscreen.Get());
+    window->SetVisible(true);
+}
+
+void GameWindowHandler::UpdateConfigFromWindow() {
+    std::vector<Recti> displays = platform->DisplayGeometries();
+
+    Pointi pos = window->Position();
+    Sizei size = window->Size();
+
+    Pointi relativePos = Pointi();
+    int display = 0;
+
+    for (size_t i = 0; i < displays.size(); i++) {
+        if (displays[i].Contains(pos)) {
+            display = i;
+            relativePos = pos - displays[i].TopLeft();
+            break;
+        }
+    }
+
+    engine->config->window.PositionX.Set(relativePos.x);
```

Comment:
Window is basically just a viewport the engine uses to position (on the display) and draw it's output.. i feel uneasy when the subordinate objects starts tweaking it's hierarchical parent's configs..
i see it's being used in CloseEvent, maybe a better approach would be for window to signal engine that close is requested? here window is simply an instrument to deliver that signal, the closing should be handled by the engine

Reply:
Yeah, agreed, this is not the best place. I've basically taped & stapled the new platform code to the existing codebase without caring much about architectural considerations because I intend to spend more time untangling the dependencies in Application / Engine / Io.

Modified it to take a config pointer, and added a TODO to move this code into a more appropriate place.